base/dates/periods.jl

```diff
@@ -166,10 +166,55 @@ type CompoundPeriod <: AbstractTime
                     break
                 end
             end
-            return new(reverse!(pc))
+            p = reverse!(pc)
+            n = length(p)
         else
             return new(resize!(p, n))
         end
+        # reduce the amount of mixed positive/negative Periods.
+        if n > 0
+            pc = sizehint!(Period[], n)
+            i = n
+            while i > 0
```

Comment:
I believe it is just O(n). Each iteration of the outer loop covers deals with a non-overlapping range. Each of those ranges is looped over twice by the interior loops.